src/log.rs

```diff
@@ -0,0 +1,471 @@
+// Copyright 2020, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use glib_sys;
+use once_cell::sync::Lazy;
+use std::boxed::Box as Box_;
+use std::sync::{Arc, Mutex};
+use translate::*;
+use GString;
+use LogLevelFlags;
+
+#[derive(Debug)]
+pub struct LogHandlerId(u32);
+
+#[doc(hidden)]
+impl FromGlib<u32> for LogHandlerId {
+    fn from_glib(value: u32) -> LogHandlerId {
+        LogHandlerId(value)
+    }
+}
+
+#[doc(hidden)]
+impl ToGlib for LogHandlerId {
+    type GlibType = u32;
+
+    fn to_glib(&self) -> u32 {
+        self.0
+    }
+}
+
+bitflags! {
+    pub struct LogFlags: u32 {
+        const FLAG_RECURSION = 1;
+        const FLAG_FATAL = 2;
+    }
+}
+
+#[doc(hidden)]
+impl ToGlib for LogFlags {
+    type GlibType = u32;
+
+    fn to_glib(&self) -> u32 {
+        self.bits()
+    }
+}
+
+#[doc(hidden)]
+impl FromGlib<glib_sys::GLogLevelFlags> for LogFlags {
+    fn from_glib(value: glib_sys::GLogLevelFlags) -> LogFlags {
+        LogFlags::from_bits_truncate(value)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum LogLevel {
+    Error,
+    Critical,
+    Warning,
+    Message,
+    Info,
+    Debug,
+}
+
+#[doc(hidden)]
+impl ToGlib for LogLevel {
+    type GlibType = u32;
+
+    fn to_glib(&self) -> u32 {
+        match *self {
+            LogLevel::Error => LogLevelFlags::LEVEL_ERROR.bits(),
+            LogLevel::Critical => LogLevelFlags::LEVEL_CRITICAL.bits(),
+            LogLevel::Warning => LogLevelFlags::LEVEL_WARNING.bits(),
+            LogLevel::Message => LogLevelFlags::LEVEL_MESSAGE.bits(),
+            LogLevel::Info => LogLevelFlags::LEVEL_INFO.bits(),
+            LogLevel::Debug => LogLevelFlags::LEVEL_DEBUG.bits(),
+        }
+    }
+}
+
+#[doc(hidden)]
+impl FromGlib<u32> for LogLevel {
+    fn from_glib(value: u32) -> LogLevel {
+        if value & LogLevelFlags::LEVEL_ERROR.bits() != 0 {
+            LogLevel::Error
+        } else if value & LogLevelFlags::LEVEL_CRITICAL.bits() != 0 {
+            LogLevel::Critical
+        } else if value & LogLevelFlags::LEVEL_WARNING.bits() != 0 {
+            LogLevel::Warning
+        } else if value & LogLevelFlags::LEVEL_MESSAGE.bits() != 0 {
+            LogLevel::Message
+        } else if value & LogLevelFlags::LEVEL_INFO.bits() != 0 {
+            LogLevel::Info
+        } else if value & LogLevelFlags::LEVEL_DEBUG.bits() != 0 {
+            LogLevel::Debug
+        } else {
+            panic!("Unknown log level: {}", value)
+        }
+    }
+}
+
+bitflags! {
+    pub struct LogLevels: u32 {
+        const LEVEL_ERROR = 4;
```

Comment:
You don't want to use the values from `glib_sys` here instead of hardcoding 4, etc?

Reply:
Oh right! It'd be much better, thanks!